### ISSUE TITLE
Auto-registering subclasses

### ISSUE THREAD

**Issue by bryant1410** [epwalsh/python-registrable#2]
I saw this repo and I think it's great to wrap up the Registrable functionality in a library.

I was wondering what are your thoughts on automatic class registration? Like, searching through the subclasses, where the name of a subclass would be the camel case version of its name.

**Comment by epwalsh** [comment]
Hey @bryant1410! I think that's a good idea. At first thought, I feel like it would be most useful if we added that functionality in a classmethod like

```python
class Registrable:
    # ... snip ...

    @classmethod
    def find_subclasses(cls, module: str) -> None:
        # recursively search module for subclasses and add them to registry
```

Is that what you had in mind?

**Comment by bryant1410** [comment]
Yeah, that could be one way, which is (pre-)loaded once, which is nice.

Another way that I thought about it was with dynamic loading, but it may be unnecessary. It'd be replacing the `_registry` dictionary with a function, like:

```python
import re
from typing import TypeVar

T = TypeVar('T', 'Registrable')


class Registrable:
    _RE_CAMEL_CASE = re.compile(r'(.)([A-Z][a-z]+)')

    @classmethod
    def name(cls) -> str:
        camel_case_name = cls.__name__[:-6]
        return cls._RE_CAMEL_CASE.sub(r'\1-\2', camel_case_name).lower()

    def _registry(cls: Type[T], name: str) -> Type[T]:
        return next(s for s in cls.__subclasses__() if s.name() == name)
```

But again, I think the use cases may be very limited to want it with dynamic loading, so I'm more inclined towards something pre-loaded. But just food for thought.

It could be nice to enable it as an option (e.g., in `__init__`), maybe with `True` as default.

The other thing I was thinking that, when pre-loading, the moment could be tricky, because all the subclasses need to be imported before. Maybe it can be detected somehow with some method so to add the new subclass to `_registry` as it's being declared. And that'd would make both looking up in the registry a constant operation and dynamic loading.

**Comment by epwalsh** [comment]
Hmm yea as far as I know it's not possible to detect subclasses without first importing the module that their defined in. But it seems like this use case is kind of already supported, i.e. by just using the fully qualified class name when calling `.by_name`: https://github.com/epwalsh/python-registrable/blob/master/registrable/__init__.py#L59. Though I think that could use some improvement. It returns anything that can be imported because it doesn't check that it is actually a subclass. So there should probably be some more safety there.

**Comment by bryant1410** [comment]
Oh, I see. I'm not convinced of using fully qualified names though cause they're too verbose.

A new subclass declaration can be detected with the class method [\_\_init_subclass\_\_](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__). In it, the direct subclass of `Registrable` can be found (e.g., if the inheritance is `Registrable -> A -> B -> C -> D` the `cls` param value would be `D` but we can register this class "as an" `A`) with a camel case version of the name (or whatever). Plus, as shown in the link, the subclass could even define a custom name optionally. Does it make sense?

**Comment by epwalsh** [comment]
Interesting. I’m still not sure what the benefit would be though, other than saving one line of code by not having to add the `@BaseClass.register(...)` decorator. But that’s not even necessary when using fully qualified names either.

**Comment by bryant1410** [comment]
It's about not having to register every new subclass that comes out as time passes, from PyTorch or wherever, but just be able to use it. They're gonna be registered by default, following a certain naming convention.

Btw, I know 3rd party library subclasses can be still be registered, it's just about practicality.

**Comment by epwalsh** [comment]
Ok, right, that would be nice to have all subclasses automatically registered when their modules are imported. But I'm not convinced that mutating the fully qualified name into something else would be a good idea. I think that if users feel like the fully qualified name is too verbose, then they could add the extra line of code to explicitly register the subclass to a name they choose.

Either way, I'm totally open to a PR with this auto-register idea using `__init_subclass` if you want to take the lead!

**Comment by bryant1410** [comment]
> Ok, right, that would be nice to have all subclasses automatically registered when their modules are imported. But I'm not convinced that mutating the fully qualified name into something else would be a good idea. I think that if users feel like the fully qualified name is too verbose, then they could add the extra line of code to explicitly register the subclass to a name they choose.

I see. I just imagine the boilerplate and I wouldn't like it :sweat_smile::

```json
  "trainer": {
    "num_epochs": 20,
    "learning_rate_scheduler": {
      "type": "torch.optim.lr_scheduler.ReduceLROnPlateau",
      "factor": 0.5,
      "mode": "max",
      "patience": 2
    },
    "optimizer": {
      "type": "torch.optim.adam.Adam",
      "betas": [0.9, 0.9]
    }
  }
```

As opposed to:

```json
  "trainer": {
    "num_epochs": 20,
    "learning_rate_scheduler": {
      "type": "reduce_lr_on_plateau",
      "factor": 0.5,
      "mode": "max",
      "patience": 2
    },
    "optimizer": {
      "type": "adam",
      "betas": [0.9, 0.9]
    }
  }
```
